content/articles/tera-template-engine.md

```diff
@@ -0,0 +1,136 @@
+Title: Introducing Tera, a template engine in Rust
+Date: 2016-04-14
+Short_summary: Talking about a template engine in Rust called Tera, inspired by Jinja2 and Django
+Category: Dev
+Authors: Vincent
+
+
+*Reading time: ~15 minutes.*
+
+Back in October 2015, I tried [Rust for web services](https://blog.wearewizards.io/trying-rust-for-web-services) and found the ecosystem lacking at the time. That's why I've been working on porting some of the tools we use in [Proppy](https://proppy.io/) to Rust: [jwt](https://crates.io/crates/jsonwebtoken), [bcrypt](https://crates.io/crates/bcrypt) (granted that Argon2 seems superior) and a [migration tool](https://crates.io/crates/dbmigrate). While I mostly do SPAs these days and don't write many templates in the backend, I still need one for some occasions. Being a Python, I like [Jinja2](http://jinja.pocoo.org/docs/dev/) and [Django templates](https://docs.djangoproject.com/en/1.9/topics/templates/#the-django-template-language). Here's how I attempted to port them to Rust and the result is [Tera](https://github.com/Keats/tera/).
+<!-- PELICAN_END_SUMMARY --> 
+
+## Goals and philosophy
+As mentioned before, the inspiration comes from both Jinja2 and django templates. As you might know, those two have similar syntax but different philosophies: Django templates are for presentation only and don't have a lot of logic allowed while Jinja2 is way more powerful.
+I side with Django on this one as complex logic is better put in code than in a template but Django goes a bit too far by even not supporting something like `{{ count + 1 }}`.
+
+So here are some of the things I want:
+
+- math operations in templates
+- no macros or other complex logic in the template
+- beautiful html output out of the box (ie no need for the `{{-` tags)
+- simple inheritance
+- simple to use filters
+- able to register new tags easily like the `{% url ... %}` in Django
+
+While new tags are definitely logic added in the template, that logic would have to be written in Rust and not in a template. That limits reusability but is simpler to understand in the end.
+
+Filters should be kept simple and be limited in scope: variable in, modifier function with optional argument and return a string. The easiest to think of would be uppercase, lowercase, capitalize and more importantly time formatting. Here are some examples of how it should look:
+```
+{{ name | uppercase }}
+{{ birthday | time:"YYYY-MM-dd" }}
+```
+Users should be able to add their own filters as well.
+
+In terms of error handling, we cannot do anything if a template cannot be parsed so panicking when encountering an error is fine and compilation should be done at compile time to ensure everything works perfectly (more on that on the README.md).
```

Comment:
"compilation should be at compile time" needs to be clearer that 1/ is template compilation and 2/ is rust compilation